R/standardize_treatment_kernel.R

```diff
@@ -0,0 +1,448 @@
+    unifw <- get_uniform_weights_treatment_kernel(nj)
+    obj <- osqp::osqp(P, q, constraints$A,
+                      constraints$l, constraints$u, pars = settings)
+    obj$WarmStart(x = unifw)
+    solution <- obj$Solve()
+  } else {
+    solution <- osqp::solve_osqp(P, q, constraints$A,
+                                 constraints$l, constraints$u,
+                                 pars = settings)
+  }
+
+  # convert weights into a matrix
+  if(verbose) message("Reordering weights...")
+  weights <- matrix(0, ncol = J, nrow = n)
+  cumsumnj <- cumsum(c(1, nj))
+  for(j in 1:J) {
+    weights[idxs[[j]], j] <- solution$x[cumsumnj[j]:(cumsumnj[j + 1] - 1)]
+  }
+
+  # compute imbalance matrix
+  imbalancetau <- as.matrix(target - t(Xtau) %*% (weights * pro_trt))
+  imbalance0 <- as.matrix(t(X0) %*% (weights * pro_trt) -
+                            t(X0) %*% (weights * pro_ctr))
+
+  # package program components if requested by user
+  if(return_program) {
+    program <- list(P = P  - lambda * I0,
+                    q = q, constraints = constraints)
+  } else {
+    program <- NULL
+  }
+
+  # return output
+  return(list(weights = weights, imbalance_0 = imbalance0,
+              imbalance_tau = imbalancetau, program = program))
+
+}
+
+#' Compute block diagonal kernel matrix
+#' @param Xs List of covariate matrices by site
+#' @param kernel Kernel to use
+compute_block_diag_kernel <- function(Xs, kernel) {
+
+  # block diagonal kernel matrix
+  kern_mat <- Matrix::bdiag(lapply(Xs, function(x) kernlab::kernelMatrix(kernel, x)))
+  return(kern_mat)
+
+}
+
+#' Create diagonal regularization matrix
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param pro_ctr_split List of J vectors of control propensity multipliers
+#' @param scale_sample_size Whether to scale the dispersion penalty by the sample size of each group, default T
+#' @param nj Number of units in each group
+#' @param n Total number of units
+create_I0_matrix_treatment_kernel <- function(pro_trt_split, pro_ctr_split, scale_sample_size, nj, n) {
+
+  if(scale_sample_size) {
+    # diagonal matrix of inverse propensity scores scaled by group sample size
+    I0 <- Matrix::Diagonal(n, rep(nj ^ 2, nj) * (unlist(pro_trt_split) + unlist(pro_ctr_split)))
+  } else {
+    # diagonal matrix of inverse propensity scores not scaled by group sample size
+    I0 <- Matrix::Diagonal(n, rep(nj, nj) * (unlist(pro_trt_split) + unlist(pro_ctr_split)))
+  }
+  return(I0)
+}
+
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param Xtaus List of J n x dtau matrices of covariates split by group
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param target Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_treatment_kernel <- function(Xtaus, pro_trt_split, target) {
+  q <- -c(do.call(rbind, Xtaus) %*% target) * unlist(pro_trt_split)
+  return(q)
+}
```

Comment:
Shouldn't this be from the kernel? Unless `kerneltau` is the linear kernel this `q` vector won't be correct. You'd have to use the kernel trick we talked about

Reply:
Good catch, thanks. I think I fixed it but not entirely sure.